static/src/javascripts/projects/common/modules/support/banner.ts

```diff
@@ -0,0 +1,205 @@
+import {
+	buildTagIds,
+	dynamicImport,
+	getArticleCountConsent,
+	isHosted,
+	ModulesVersion,
+	supportDotcomComponentsUrl,
+	tracking,
+} from 'common/modules/support/supportMessaging';
+import type { Page } from 'common/modules/support/supportMessaging';
+import userPrefs from 'common/modules/user-prefs';
+import config from 'lib/config';
+import fastdom from 'lib/fastdom-promise';
+import { getCountryCode } from 'lib/geolocation';
+import reportError from 'lib/report-error';
+
+export const NO_RR_BANNER_TIMESTAMP_KEY = 'gu.noRRBannerTimestamp'; // timestamp of when we were last told not to show a RR banner
+const twentyMins = 20 * 60000;
+
+export const withinLocalNoBannerCachePeriod = (): boolean => {
+	const item = window.localStorage.getItem(NO_RR_BANNER_TIMESTAMP_KEY);
+	if (item && !Number.isNaN(parseInt(item, 10))) {
+		const withinCachePeriod = parseInt(item, 10) + twentyMins > Date.now();
+		if (!withinCachePeriod) {
+			// Expired
+			window.localStorage.removeItem(NO_RR_BANNER_TIMESTAMP_KEY);
+		}
+		return withinCachePeriod;
+	}
+	return false;
+};
+
+export const setLocalNoBannerCachePeriod = (): void =>
+	window.localStorage.setItem(NO_RR_BANNER_TIMESTAMP_KEY, `${Date.now()}`);
+
+export const renderBanner = (
+	response: ModuleDataResponse,
+): Promise<boolean> => {
+	if (!response.data) {
+		return Promise.resolve(false);
+	}
+	const { module, meta } = response.data;
+
+	return dynamicImport(module.url, module.name)
+		.then((Banner: React.FC<any>) => {
+			const isPuzzlesBanner = module.name === 'PuzzlesBanner';
+
+			return fastdom
+				.mutate(() => {
+					const container = document.createElement('div');
+					container.classList.add('site-message--banner');
+					container.classList.add('remote-banner');
+					if (isPuzzlesBanner) {
+						container.classList.add('remote-banner--puzzles');
+					}
+
+					document.body.insertAdjacentElement('beforeend', container);
+
+					return mountDynamic(
+						container,
+						Banner,
+						{ submitComponentEvent, ...module.props },
+						!isPuzzlesBanner, // The puzzles banner has its own CacheProvider component, and needs this to be false
+					);
+				})
+				.then(() => {
+					// track banner view event in Google Analytics for subscriptions banner
+					if (
+						meta.componentType ===
+						'ACQUISITIONS_SUBSCRIPTIONS_BANNER'
+					) {
+						trackNonClickInteraction(
+							'subscription-banner : display',
+						);
+					}
+
+					return true;
+				});
+		})
+		.catch((error: any) => {
+			/* eslint-disable @typescript-eslint/restrict-template-expressions -- error log */
+			console.log(`Error importing remote banner: ${error}`);
+			reportError(
+				new Error(`Error importing remote banner: ${error}`),
+				{},
+				false,
+			);
+			/* eslint-enable @typescript-eslint/restrict-template-expressions */
+			return false;
+		});
+};
+
+const buildBannerPayload = async (): Promise<BannerPayload> => {
+	const page = config.get('page') as Page;
```

Comment:
There's two options to make this type-safe.

Either you use type suggestions for config:

```suggestion
	const page = config.get<Page>('page', {});
```

The preferred approach is to access the window.guardian.page object directly, and updating the type declaration accordingly. This is preferable as it can help us identify exactly what is actually used by our code, and possibly trim the page object in the future.

cc @guardian/dotcom-platform

Reply:
ahh I didn't realise we have typed `window.guardian.config.page`, that is very useful

Reply:
I've updated it to use `window.guardian.config.page`